src/overview.md

```diff
@@ -0,0 +1,322 @@
+program) is [`rustc::ty::Ty`][ty]. This is so important that we have a whole chapter
+on [`ty::Ty`][ty], but for now, we just want to mention that it exists and is the way
+`rustc` represents types!
+
+Oh, and also the `rustc::ty` module defines the `TyCtxt` struct we mentioned before.
+
+[ty]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/type.Ty.html
+
+### Parallelism
+
+Compiler performance is a problem that we would like to improve on
+(and are always working on). One aspect of that is parallelizing
+`rustc` itself.
+
+Currently, there is only one part of rustc that is already parallel: codegen.
+During monomorphization, the compiler will split up all the code to be
+generated into smaller chunks called _codegen units_. These are then generated
+by independent instances of LLVM. Since they are independent, we can run them
+in parallel. At the end, the linker is run to combine all the codegen units
+together into one binary.
+
+However, the rest of the compiler is still not yet parallel. There have been
+lots of efforts spent on this, but it is generally a hard problem. The current
+approach is (**TODO: verify**) to turn `RefCell`s into `Mutex`s -- that is, we
+switch to thread-safe internal mutability. However, there are ongoing
+challenges with lock contention, maintaining query-system invariants under
+concurrency, and the complexity of the code base. One can try out the current
+work by enabling parallel compilation in `config.toml`. It's still early days,
+but there are already some promising performance improvements.
+
+### Bootstrapping
+
+**TODO (or do we want such a section)?**
+
+# Unresolved Questions
+
+**TODO: find answers to these**
+
+- Does LLVM ever do optimizations in debug builds?
+- How do I explore phases of the compile process in my own sources (lexer,
+  parser, HIR, etc)? - e.g., `cargo rustc -- -Zunpretty=hir-tree` allows you to
+  view HIR representation
+- What is the main source entry point for `X`?
+- Where do phases diverge for cross-compilation to machine  code across
+  different platforms?
+
+# References
+
+- Command line parsing
+    - Guide: [The Rustc Driver and Interface](https://rust-lang.github.io/rustc-guide/rustc-driver.html)
+    - Driver definition: [`rustc_driver`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_driver/)
+    - Main entry point: **TODO**
+- Lexical Analysis: Lex the user program to a stream of tokens
+    - Guide: [Lexing and Parsing](https://rust-lang.github.io/rustc-guide/the-parser.html)
+    - Lexer definition: [`librustc_lexer`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lexer/index.html)
+    - Main entry point: **TODO**
+- Parsing: Parse the stream of tokens to an Abstract Syntax Tree (AST)
+    - Guide: [Lexing and Parsing](https://rust-lang.github.io/rustc-guide/the-parser.html)
+    - Parser definition: [`librustc_parse`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_parse/index.html)
+    - Main entry point: **TODO**
```

Comment:
I believe this is [`rustc_parse::parse_in`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_parse/fn.parse_in.html).
```suggestion
    - Main entry point: [`rustc_parse::parse_in`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_parse/fn.parse_in.html)
```

Reply:
No, that's a function that's primarily used for parsing the `TokenStream` inside an `Attribute`, and not for general parsing.

Reply:
@Centril Can you point to the correct entry point?

Reply:
Well there are a few entry points depending on the situation. For example, the lib.rs file has a different entry point (called by a query in the rustc_interface) than parsing an outline module (called by expansion), and then there are black box parsers for macro fragments. Which do you want?

Reply:
Preferably all of them if possible.

Reply:
Does `rustc_interface` call the parser which calls the lexer?  This seems to (maybe) be the case and is what I am expecting to find but I am having a tough time locating the connections beginning at `rustc_interface`.

Reply:
> Preferably all of them if possible.

- Entry point for first file in crate is: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_interface/passes/fn.parse.html

- Entry point for outline module parsing is: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_expand/module/fn.parse_external_mod.html

- Entry point for macro fragments is: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_expand/mbe/macro_parser/fn.parse_nt.html (I might move this into the parser to reduce visibilities there.)

- There are more, but these are the most notable ones.

These lead into the parser crate itself, but all of them are outside.

> Does `rustc_interface` call the parser which calls the lexer? This seems to (maybe) be the case and is what I am expecting to find but I am having a tough time locating the connections beginning at `rustc_interface`.

The interface calls a glue function between the lexer and the parser that composes them together for convenience.

Reply:
I added Mazdak's links in 5243c7a3f7bd1d0a7b85ee4247219f5581ce2e30 (#669)

Reply:
I think that I have the first file in crate full call path worked out from `rustc_interface` -> `parser` entry point, including how the lexer is instantiated and what is called to return a `TokenStream` (it appears to be part of the `Parser` instantiation).  

I'll summarize this for the Overview in #669 and will add more detailed documentation in the Guide chapter.

Reply:
@Centril I added an initial draft of Overview parser documentation in  https://github.com/rust-lang/rustc-dev-guide/pull/669/commits/79c19ab99818a9184e037dad0bb496eac4c0292f

This does not include information about macro expansion, AST validation, nameres, or early linting yet.  How detailed do you feel that we should dive into these topics in the overview?  Mention that they occur and then expand on them in the Guide chapter?

Reply:
@chrissimpkins I'd just include the entry point in the overview and then call it a day; most of the additional stuff belongs in a page about the parser imo.

Reply:
> @chrissimpkins I'd just include the entry point in the overview and then call it a day; most of the additional stuff belongs in a page about the parser imo.

Thank you! I will update it today and begin to move the other information over to the Guide chapter.

Reply:
I think that I addressed this as you recommended in https://github.com/rust-lang/rustc-dev-guide/pull/669/commits/b3e75e8b7b32c71e7d97f2e98a262ea279cfed27

Reply:
@chrissimpkins I cherry-picked the commit with the entry points from your branch

Reply:
> @chrissimpkins I cherry-picked the commit with the entry points from your branch

Ty!